### PR TITLE
Raise event in case using default Slirp binding plugin image 

### PR DESCRIPTION
**Pull request by ormergi**
**What this PR does / why we need it**:
Raise verbosity of warning by rasing an event  in case no Slirp network binding plugin image is registered and Kubevirt use the default image.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
It was necessary to pass the VMI event recorder to the template service.

Handling Slirp default image is moved from `ReadNetBindingPluginConfiguration` to the caller because an raising an event requires the VMI event recorder and the VMI objects.

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by ormergi** [comment]
One test failed since the assertion fail if a warning is raised, which is expected because the warning is expacted to fail in that case 
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/10404/pull-kubevirt-e2e-k8s-1.27-sig-network/1700872711619743744

File: tests/testsuite/runconfiguration.go [hunk 1]
```diff
@@ -23,5 +23,20 @@ func initRunConfiguration(virtClient kubecli.KubevirtClient) {
 		runConfig.WarningToIgnoreList = append(runConfig.WarningToIgnoreList, "EvictionStrategy is set but vmi is not migratable")
 	}
 
+	if !hasRegisteredSlirpBindingPluginImage(kv) {
```
Comment (AlonaKaplan) [inline tests/testsuite/runconfiguration.go:26]: Instead of ignoring it, I would add to the relevant tests an expect that expects getting the warning.

**Comment by EdDev** [comment]
Here is a suggestion to extract and decouple the current sidecar creation definition logic out of the template rendering: https://github.com/kubevirt/kubevirt/pull/10423

I think it should also assist in avoiding adding more code and dependency (on the event).

File: tests/testsuite/runconfiguration.go (hunk 1 above)
Comment (ormergi) [reply tests/testsuite/runconfiguration.go]: Done

**Comment by ormergi** [comment]
@EdDev @AlonaKaplan Please take another look

**Comment by ormergi** [comment]
> Here is a suggestion to extract and decouple the current sidecar creation definition logic out of the template > rendering: https://github.com/kubevirt/kubevirt/pull/10423

> I think it should also assist in avoiding adding more code and dependency (on the event).

Thanks for suggesting this I will go over it as soon as I can, I prefer to not block this PR on it.

File: pkg/virt-controller/services/template.go [hunk 2]
```diff
@@ -372,14 +372,20 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, i
 		return i.Slirp != nil
 	})
 	if len(slirpIfaces) > 0 {
+		var slirpSidecarImage string
 		if plugin := ReadNetBindingPluginConfiguration(t.clusterConfig.GetConfig(), SlirpNetworkBindingPluginName); plugin == nil {
-			return nil, fmt.Errorf("couldn't find %s network binding plugin configuration, make sure its specified in Kubevirt config", SlirpNetworkBindingPluginName)
+			// In case no Slirp network binding plugin is registered (i.e.: specified in in Kubevirt config) use default image
```
Comment (AlonaKaplan) [inline pkg/virt-controller/services/template.go:377]: This file is long enough, I prefer to keep this logic in the network specific file (even if it means the event manager has to be passed one step deeper).

**Comment by ormergi** [comment]
[Changes](https://github.com/kubevirt/kubevirt/compare/195ed52afa26c95c19edde04733ef44a9d7bda70..0bb2b4a3b910f30f44ec94d4a9a426fd80c95b8b): 
- Change template service factory to receive options for the vmi recoreder, this way changing the factory signature doesnt affect all the places it being used. 
- Export Slirp hook creation to func and move setting the default image to it, in order to simplify the implementation and reduce abstraction layers.

File: pkg/virt-controller/services/template.go (hunk 2 above)
Comment (ormergi) [reply pkg/virt-controller/services/template.go]: Done

File: pkg/virt-controller/services/net_binding_test.go [hunk 3]
```diff
@@ -103,28 +107,34 @@ var _ = Describe("Network Binding", func() {
...
+			}))
 		})
 
-		DescribeTable("should return no network binding plugins configuration, when config", func(config *v1.KubeVirtConfiguration) {
-			Expect(services.ReadNetBindingPluginConfiguration(config, "myplugin")).To(BeNil())
-		},
-			Entry("is nil", nil),
+		DescribeTable("should create Slirp hook sidecar with default image, when Kubevirt config",
+			func(config *v1.KubeVirtConfiguration) {
+				fakeRecorder := record.NewFakeRecorder(1)
```
Comment (EdDev) [inline pkg/virt-controller/services/net_binding_test.go:131]: I think you are suppose to check that the event was emitted by examining `fakeRecorder.Events`.

File: tests/network/vmi_networking.go [hunk 4]
```diff
@@ -415,7 +415,10 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
 			deadbeafVMI, err = virtClient.VirtualMachineInstance(testsuite.GetTestNamespace(nil)).Create(context.Background(), deadbeafVMI)
 			Expect(err).ToNot(HaveOccurred())
 
-			libwait.WaitUntilVMIReady(deadbeafVMI, console.LoginToAlpine)
+			const unregisterSlipImageWarning = "no Slirp network binding plugin image is set in Kubevirt config, using " +
```
Comment (EdDev) [inline tests/network/vmi_networking.go:418]: This feels very fragile to me, any change to the text will break this test.
I am not sure if it is worth using this ignore list or just ignore everything.

**Review by EdDev — COMMENTED** [review]
Thanks, this version seems less intrusive than the last one.
I would still prefer not to add that event handler object into the template service object, but at least this is easier to revert.

While I still think this needs to be rebased on the decoupling PR proposed, I will not block on this. I can fix it after by reverting the dependency, although it is ugly.

File: pkg/virt-controller/services/template.go [hunk 5]
```diff
@@ -372,14 +372,7 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, i
...
 	if len(slirpIfaces) > 0 {
-		if plugin := ReadNetBindingPluginConfiguration(t.clusterConfig.GetConfig(), SlirpNetworkBindingPluginName); plugin == nil {
-			return nil, fmt.Errorf("couldn't find %s network binding plugin configuration, make sure its specified in Kubevirt config", SlirpNetworkBindingPluginName)
-		} else {
-			requestedHookSidecarList = append(requestedHookSidecarList, hooks.HookSidecar{
-				Image:           plugin.SidecarImage,
-				ImagePullPolicy: t.clusterConfig.GetImagePullPolicy(),
-			})
-		}
+		requestedHookSidecarList = append(requestedHookSidecarList, SlirpNetBindingPluginSidecar(vmi, t.clusterConfig.GetConfig(), t.recorder))
```
Comment (AlonaKaplan) [inline pkg/virt-controller/services/template.go:375]: Can you please move all the slirp related code to `NetBindingPluginSidecarList`? So the template file won't have any of this logic and all the binding sidecar logic will reside in the net_binding file.

Comment (ormergi) [reply pkg/virt-controller/services/template.go:375]: Note that `SlirpNetBindingPluginSidecar` now requires the event recorder, it will require changing `NetBindingPluginSidecarList ` to receive it, WDYT?

File: pkg/virt-controller/services/net_binding_test.go (hunk 3 above)
Comment (ormergi) [reply pkg/virt-controller/services/net_binding_test.go:131]: Do we check warning events emitting in unit tests?

File: pkg/virt-controller/services/template.go (hunk 5 above)
Comment (ormergi) [reply pkg/virt-controller/services/template.go]: Done

File: pkg/virt-controller/services/net_binding.go [hunk 6]
```diff
@@ -44,6 +45,8 @@ func NetBindingPluginSidecarList(vmi *v1.VirtualMachineInstance, config *v1.Kube
 			if !exist {
 				return nil, fmt.Errorf("couldn't find configuration for network bindining: %s", iface.Binding.Name)
 			}
+		} else if iface.Slirp != nil {
+			pluginSidecars = append(pluginSidecars, SlirpNetBindingPluginSidecar(vmi, config, recorder))
 		}
 	}
 	for _, pluginInfo := range bindingByName {
```
Comment (AlonaKaplan) [inline pkg/virt-controller/services/net_binding.go:52]: For the consistency, I would add `ImagePullPolicy: kvConfig.ImagePullPolicy` to this image as well.

File: pkg/virt-controller/services/net_binding_test.go [hunk 7]
```diff
@@ -98,33 +102,39 @@ var _ = Describe("Network Binding", func() {
...
 				NetworkConfiguration: &v1.NetworkConfiguration{},
 			}
-			_, err := services.NetBindingPluginSidecarList(vmi, config)
+			_, err := services.NetBindingPluginSidecarList(vmi, config, record.NewFakeRecorder(1))
 			Expect(err).To(HaveOccurred())
 		})
 	})
 
-	Context("binding plugin configuration", func() {
-		It("should read registered network binding plugins successfully", func() {
```
Comment (AlonaKaplan) [inline pkg/virt-controller/services/net_binding_test.go:107]: I don't understand why the original tests were removed. Where are the unit tests checking non sriov plugins or SRIOV as a plugin (and not the old API flow).

File: tests/network/vmi_networking.go [hunk 8]
```diff
@@ -415,7 +415,10 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
...
+				"'quay.io/kubevirt/network-slirp-binding:20230830_638c60fc8' sidecar image for Slirp network binding configuration"
+			warnings := append(testsuite.TestRunConfiguration.WarningToIgnoreList, unregisterSlipImageWarning)
+			libwait.WaitUntilVMIReady(deadbeafVMI, console.LoginToAlpine, libwait.WithWarningsIgnoreList(warnings))
```
Comment (AlonaKaplan) [inline tests/network/vmi_networking.go:421]: I think the other way around should be done (I"m pretty sure I already wrote this comment, but I cannot find it). We should verify the warning is printed instead of ignoring it. That's the new logic of this PR.

File: pkg/virt-controller/services/net_binding_test.go (hunk 7 above)
Comment (ormergi) [reply pkg/virt-controller/services/net_binding_test.go:107]: Sorry, I don't follow, this commit adapt unit tests following the removal of `ReadNetBindingPluginConfiguration` in favor of  `SlirpNetBindingPluginSidecar`

File: pkg/virt-controller/services/net_binding.go (hunk 6 above)
Comment (ormergi) [reply pkg/virt-controller/services/net_binding.go:52]: Done

File: tests/network/vmi_networking.go (hunk 8 above)
Comment (AlonaKaplan) [reply tests/network/vmi_networking.go:421]: ?

**Comment by AlonaKaplan** [comment]
/approve
Thanks!

File: pkg/virt-controller/services/template.go [hunk 9]
```diff
@@ -1137,6 +1139,8 @@ func getNamespaceAndNetworkName(namespace string, fullNetworkName string) (strin
 	return precond.MustNotBeEmpty(namespace), fullNetworkName
 }
 
+type TemplateServiceOption func(service *templateService)
```
Comment (EdDev) [inline pkg/virt-controller/services/template.go:1127]: There is no need to expose it, `templateServiceOption` should be enough.
There is also no need to name the argument of the function, the type is informative enough.

File: pkg/virt-controller/services/template.go [hunk 10]
```diff
@@ -1149,7 +1153,8 @@ func NewTemplateService(launcherImage string,
 	virtClient kubecli.KubevirtClient,
 	clusterConfig *virtconfig.ClusterConfig,
 	launcherSubGid int64,
-	exporterImage string) TemplateService {
+	exporterImage string,
+	opts ...TemplateServiceOption) TemplateService {
```
Comment (EdDev) [inline pkg/virt-controller/services/template.go:1142]: nit: Please move the return type to its own separate line like [here](https://github.com/kubevirt/kubevirt/pull/10423/files#diff-e2b17b934fcd4fb2278da4c106432d96e03f936e34e2dd4854843501aa719ad8R1143).

Every time someone adds something the diff shows too much.

File: pkg/virt-controller/watch/application.go [hunk 11]
```diff
@@ -603,6 +603,7 @@ func (vca *VirtControllerApp) initCommon() {
 		vca.clusterConfig,
 		vca.launcherSubGid,
 		vca.exporterImage,
+		services.WithEventRecorder(vca.vmiRecorder),
```
Comment (EdDev) [inline pkg/virt-controller/watch/application.go:606]: Do we agree on removing this in a follow up?
In favor of extracting the sidecar registration? (e.g. https://github.com/kubevirt/kubevirt/pull/10423).

File: pkg/virt-controller/services/net_binding.go [hunk 12]
```diff
@@ -44,33 +45,56 @@ func NetBindingPluginSidecarList(vmi *v1.VirtualMachineInstance, config *v1.Kube
 			if !exist {
 				return nil, fmt.Errorf("couldn't find configuration for network bindining: %s", iface.Binding.Name)
 			}
+		} else if iface.Slirp != nil {
+			pluginSidecars = append(pluginSidecars, SlirpNetBindingPluginSidecar(vmi, config, recorder))
```
Comment (EdDev) [inline pkg/virt-controller/services/net_binding.go:49]: Before we had this code block creating a `bindingByName` pool and only later appending to the sidecars pool. Now we have a mix that makes me dizzy.

Just do another for-loop and do not mix things together, it makes it hard to follow this and hard to move later.

File: pkg/virt-controller/services/net_binding_test.go (hunk 3 above)
Comment (EdDev) [reply pkg/virt-controller/services/net_binding_test.go:131]: You have a flake here, you are suppose to check what it recorded. That is why a fake was created, to allow such checks.

File: tests/network/vmi_networking.go (hunk 8 above)
Comment (EdDev) [reply tests/network/vmi_networking.go:422]: We could be fine with making sure it is tested in unit tests.
Checking events in e2e here is borderline to me, it somehow relates a bit to checking logs.

But if it is easy to do, I guess no harm in having it.

File: pkg/virt-controller/services/template.go (hunk 9 above)
Comment (ormergi) [reply pkg/virt-controller/services/template.go:1127]: Done

File: pkg/virt-controller/services/template.go (hunk 10 above)
Comment (ormergi) [reply pkg/virt-controller/services/template.go:1142]: Done

File: pkg/virt-controller/services/net_binding.go (hunk 12 above)
Comment (ormergi) [reply pkg/virt-controller/services/net_binding.go:49]: Done

File: pkg/virt-controller/services/net_binding_test.go (hunk 3 above)
Comment (ormergi) [reply pkg/virt-controller/services/net_binding_test.go:131]: A fake was created just to satisfy `services.SlirpNetBindingPluginSidecar`, this tests are about checking slirp default image is set.
No flakes whatsoever.

I added separate test to verify an event is raised.

File: pkg/virt-controller/watch/application.go (hunk 11 above)
Comment (ormergi) [reply pkg/virt-controller/watch/application.go:606]: Fine by me.

File: pkg/virt-controller/services/net_binding.go [hunk 13]
```diff
@@ -46,31 +47,60 @@ func NetBindingPluginSidecarList(vmi *v1.VirtualMachineInstance, config *v1.Kube
 			}
 		}
 	}
+
+	for _, iface := range vmi.Spec.Domain.Devices.Interfaces {
+		if iface.Slirp != nil {
+			pluginSidecars = append(pluginSidecars, SlirpNetBindingPluginSidecar(vmi, config, recorder))
+			break
```
Comment (EdDev) [inline pkg/virt-controller/services/net_binding.go:54]: I do not think this is needed. Embedding here the logic that there can be only one such interface is not a fit IMO.

File: pkg/virt-controller/services/net_binding_test.go [hunk 14]
```diff
@@ -98,33 +102,39 @@ var _ = Describe("Network Binding", func() {
...
-			Expect(services.ReadNetBindingPluginConfiguration(config, "myplugin")).To(BeNil())
-		},
-			Entry("is nil", nil),
+		DescribeTable("should create Slirp hook sidecar with default image, when Kubevirt config",
+			func(config *v1.KubeVirtConfiguration) {
+				fakeRecorder := record.NewFakeRecorder(1)
+				vmi := v1.NewVMI("test", "1234")
+
+				Expect(services.SlirpNetBindingPluginSidecar(vmi, config, fakeRecorder)).To(Equal(hooks.HookSidecar{
+					Image: "quay.io/kubevirt/network-slirp-binding:20230830_638c60fc8",
```
Comment (EdDev) [inline pkg/virt-controller/services/net_binding_test.go:135]: I think you should use `DefaultSlirpPluginImage` to express this is a specific fixed one.

File: pkg/virt-controller/services/net_binding_test.go [hunk 15]
```diff
@@ -149,9 +159,22 @@ var _ = Describe("Network Binding", func() {
...
 
-		It("should return default image when no Slirp network binding plugin image is registered", func() {
-			expectedIfaceBindingPlugin := &v1.InterfaceBindingPlugin{SidecarImage: "quay.io/kubevirt/network-slirp-binding:20230830_638c60fc8"}
-			Expect(services.ReadNetBindingPluginConfiguration(&v1.KubeVirtConfiguration{}, "slirp")).To(Equal(expectedIfaceBindingPlugin))
+		It("should raise UnregisteredNetworkBindingPlugin warning event when no slirp binding plugin image is registered (specified in Kubevirt config)", func() {
+			fakeRecorder := record.NewFakeRecorder(1)
+			vmi := v1.NewVMI("test", "1234")
+
+			Expect(services.SlirpNetBindingPluginSidecar(vmi, &v1.KubeVirtConfiguration{}, fakeRecorder)).To(Equal(hooks.HookSidecar{
+				Image: "quay.io/kubevirt/network-slirp-binding:20230830_638c60fc8",
```
Comment (EdDev) [inline pkg/virt-controller/services/net_binding_test.go:167]: I think you should use `DefaultSlirpPluginImage` to express this is the default one.

File: pkg/virt-controller/services/net_binding_test.go [hunk 16]
```diff
@@ -149,9 +159,22 @@ var _ = Describe("Network Binding", func() {
...
+			}))
+
+			select {
```
Comment (EdDev) [inline pkg/virt-controller/services/net_binding_test.go:170]: Why do you use `select` here?
You could just assert that there is one event (`len(fakeRecorder.Events) ==  1`) and then fetch & process it.

File: pkg/virt-controller/services/net_binding_test.go [hunk 17]
```diff
@@ -149,9 +159,22 @@ var _ = Describe("Network Binding", func() {
...
+			case event := <-fakeRecorder.Events:
+				Expect(event).To(Equal(
+					"Warning " +
+						"UnregisteredNetworkBindingPlugin " +
+						"no Slirp network binding plugin image is set in Kubevirt config, using " +
+						"'quay.io/kubevirt/network-slirp-binding:20230830_638c60fc8' sidecar image for Slirp network binding configuration"))
```
Comment (EdDev) [inline pkg/virt-controller/services/net_binding_test.go:176]: Could you please allign the lines and use the constants for the image and event ID?

File: pkg/virt-controller/services/net_binding_test.go (hunk 16 above)
Comment (ormergi) [reply pkg/virt-controller/services/net_binding_test.go:170]: Since we define and expect to single event both ways works, I dont mind changing it to what you suggest.
Done.

File: pkg/virt-controller/services/net_binding_test.go (hunk 17 above)
Comment (ormergi) [reply pkg/virt-controller/services/net_binding_test.go:176]: Done

File: pkg/virt-controller/services/net_binding_test.go (hunk 15 above)
Comment (ormergi) [reply pkg/virt-controller/services/net_binding_test.go:167]: Done

File: pkg/virt-controller/services/net_binding_test.go (hunk 14 above)
Comment (ormergi) [reply pkg/virt-controller/services/net_binding_test.go:135]: Done
